sbncode/EventGenerator/MultiPart/MultiPartRain_module.cc

```diff
@@ -0,0 +1,464 @@
+// from cetpkgsupport v1_11_00.
+////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+//#include "art/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+#include <memory>
+#include <vector>
+
+#include "CLHEP/Random/RandFlat.h"
+#include "TRandom.h"
+#include "nurandom/RandomUtils/NuRandomService.h"
+#include "larcore/Geometry/Geometry.h"
+#include "larcoreobj/SummaryData/RunData.h"
+
+#include "nusimdata/SimulationBase/MCTruth.h"
+#include "nusimdata/SimulationBase/MCParticle.h"
+
+#include "TLorentzVector.h"
+#include "TDatabasePDG.h"
+
+struct PartGenParam {
+    std::vector<int       > pdg;
+    std::vector<double    > mass;
+    std::array <size_t, 2 > multi;
+    std::array <double, 2 > kerange;
+    bool use_mom;
+    bool direct_inward;
+    double weight;
+};
+
+class MultiPartRain;
+
+class MultiPartRain : public art::EDProducer {
+public:
+  explicit MultiPartRain(fhicl::ParameterSet const & p);
+  // The destructor generated by the compiler is fine for classes
+  // without bare pointers or other resource use.
+  ~MultiPartRain();
+  
+  // Plugins should not be copied or assigned.
+  MultiPartRain(MultiPartRain const &) = delete;
+  MultiPartRain(MultiPartRain &&) = delete;
+  MultiPartRain & operator = (MultiPartRain const &) = delete;
+  MultiPartRain & operator = (MultiPartRain &&) = delete;
+  
+  // Required functions.
+  void produce(art::Event & e) override;
+  
+  void beginRun(art::Run& run) override;
+  
+  void GenPosition(double& x, double& y, double& z);
+  
+  void GenMomentum(const PartGenParam& param, const double& mass, double& px, double& py, double& pz,
+		   const double x, const double y, const double z);
+    
+    std::vector<size_t> GenParticles() const;
+    
+private:
+    
+    CLHEP::HepRandomEngine& fFlatEngine;
+    CLHEP::RandFlat *fFlatRandom;
+    
+    // exception thrower
+    void abort(const std::string msg) const;
+    
+    // array of particle info for generation
+    std::vector<PartGenParam> _param_v;
+    
+    // g4 time of generation
+    double _t0;
+    double _t0_sigma;
+    
+    // g4 position
+    std::array<double,2> _xrange;
+    std::array<double,2> _yrange;
+    std::array<double,2> _zrange;
+
+    // TPC array
+    std::vector<std::vector<unsigned short> > _tpc_v;
+    
+    // multiplicity constraint
+    size_t _multi_min;
+    size_t _multi_max;
+    
+    // verbosity flag
+    unsigned short _debug;
+};
+
+void MultiPartRain::abort(const std::string msg) const
+{
+    std::cerr << "\033[93m" << msg.c_str() << "\033[00m" << std::endl;
+    throw std::exception();
```

Comment:
I would prefer to avoid throwing empty exception as they are not very informative. but ok.